client/src/main/java/com/paypal/selion/platform/grid/SeleniumGridListener.java

```diff
@@ -137,10 +137,13 @@ public void beforeInvocation(IInvokedMethod method, ITestResult testResult) {
                     System.exit(1); // NOSONAR
                 }
             }
+        } catch (RuntimeException e) { // NOSONAR
+            Grid.getThreadLocalException().set(e);
         } catch (Exception e) { // NOSONAR
-            logger.log(Level.WARNING, "An error occurred while processing beforeInvocation: " + e.getMessage(), e);
+            //Must wrap the Exception to RuntimeException, So that users no need to
+            //add try catch block in their test case when invoke Grid.driver()
+            Grid.getThreadLocalException().set(new RuntimeException(e.getMessage(), e));
```

Comment:
What I was recommending for your recent addition was to also start using this instead:
private static ThreadLocal<Exception> threadLocalException = new ThreadLocal<Exception>();

I thought that in the sharing case we could have several exceptions in beforeX but with this recent addition in beforeInvocation:

if (Grid.getTestSession() != null) {
                return;
 }

That is not possible now.

So ignore my 2nd comment: